frappe/public/js/frappe/utils/utils.js

```diff
@@ -1112,75 +1112,104 @@ Object.assign(frappe.utils, {
+					value +
+						__(
+							this.duration_data[unit].abbr,
+							null,
+							unit.toUpperCase() + " (Field: Duration)"
+						)
+				);
 			}
-			if (total_duration.seconds) {
-				duration += duration.length ? " " : "";
-				duration += total_duration.seconds + __("s", null, "Seconds (Field: Duration)");
+		}
+		return components.join(" ");
+	},
+
+	/**
+	 * Turn seconds into a duration object.
+	 *
+	 * @param {number} seconds - Duration in seconds
+	 * @param {Object} [duration_options] - Formatting options
+	 * 	Example: { hide_days: true, hide_months: true }
+	 * @returns {Object} - Duration object with each part not hidden
+	 */
+	seconds_to_duration: function (seconds, duration_options = null) {
+		let duration = {};
+		let remaining_seconds = seconds;
+		for (const [unit, { factor }] of Object.entries(this.duration_data)) {
+			if (!duration_options?.[`hide_${unit}`]) {
+				duration[unit] = Math.floor(remaining_seconds / factor);
+				remaining_seconds %= factor;
 			}
 		}
 		return duration;
 	},
 
-	seconds_to_duration(seconds, duration_options) {
-		const round = seconds > 0 ? Math.floor : Math.ceil;
-		const total_duration = {
-			days: round(seconds / 86400), // 60 * 60 * 24
-			hours: round((seconds % 86400) / 3600),
-			minutes: round((seconds % 3600) / 60),
-			seconds: round(seconds % 60),
-		};
-
-		if (duration_options && duration_options.hide_days) {
-			total_duration.hours = round(seconds / 3600);
-			total_duration.days = 0;
+	/**
+	 * Converts various time units to seconds.
+	 *
+	 * @param {Object} units - The time units to convert.
+	 * @returns {number} - The total time in seconds.
+	 */
+	duration_to_seconds: function (units = {}) {
+		let seconds = 0;
+		for (const [unit, { factor }] of Object.entries(this.duration_data)) {
+			seconds += (units[unit] || 0) * factor;
 		}
-
-		return total_duration;
+		return seconds;
 	},
 
-	duration_to_seconds(days = 0, hours = 0, minutes = 0, seconds = 0) {
-		let value = 0;
-		if (days) {
-			value += days * 24 * 60 * 60;
-		}
-		if (hours) {
-			value += hours * 60 * 60;
-		}
-		if (minutes) {
-			value += minutes * 60;
-		}
-		if (seconds) {
-			value += seconds;
+	/**
+	 * Retrieves duration options from a given docfield.
+	 *
+	 * @param {Object} docfield - The docfield to get options from.
+	 * @returns {Object} - The duration options.
+	 */
+	get_duration_options: function (docfield) {
+		const options = {};
+		for (const option of Object.keys(this.duration_data)) {
+			const hideOption = `hide_${option}`;
+			if (hideOption in docfield) {
+				options[hideOption] = docfield[hideOption];
+			}
 		}
-		return value;
+		return options;
 	},
 
-	get_duration_options: function (docfield) {
-		return {
-			hide_days: docfield.hide_days,
-			hide_seconds: docfield.hide_seconds,
-		};
+	/**
+	 * Returns all defined duration parts.
+	 *
+	 * @returns {Array} - The list of duration parts.
+	 */
+	get_duration_data: function () {
+		return Object.keys(this.duration_data);
```

Comment:
Helper for implementing code.
Could be left out, is only used once and rather trivial, as units don’t change.